### PR TITLE
Add term: Debugging

### PR DESCRIPTION
**Pull request by fsbaraglia**

### CODE REVIEW

File: content/en/debugging.md
Comment (CathPag) [inline content/en/debugging.md:14]:
```suggestion
Debugging tools use various debugging techniques to help developers identify bugs. How these tools are used depends on the development stage. Analysis of logs, traces, and metrics, for instance, are used for debugging directly in production without requiring program execution to stop. Developers use interactive debugging on the source code during the programming phase. By running the program and connecting an external debugger tool, developers control the flow of the program and instruct it to pause the execution at specific points where they suspect the error may happen; they usually get this information by reading the logs. Once they have identified the source of the failure, they can correct the code and create a bug fix or patch.
```

Comment (CathPag) [reply content/en/debugging.md:14]: Just edits

File: content/en/debugging.md
Comment (CathPag) [inline content/en/debugging.md:3]: I changed it to feedback appreciated as it still needs to be reviewed by another approver

**Review by CathPag — COMMENTED** [review]
Thanks, @fsbaraglia!  Looks good to me. Submitted a few editorial changes and have one question for the 2nd section.

File: content/en/debugging.md
Comment (fsbaraglia) [reply content/en/debugging.md]: yes, you got perfectly the point :D I was missing some words, this is the reason that I submitted to get some ideas from you :D thank you very much

File: content/en/debugging.md
Comment (CathPag) [inline content/en/debugging.md:11]: Just realized one more thing, sorry. Here we determined that the problem is that we don't have any context to debug an app manually.

Presumably, debugging tools solve that by providing that needed context but we don't mention it below...or at least it's not clear how that relates to things like variables, data, location in memory, etc. 

How it helps should directly address the problem. It's probably already there but implied. We should spell it out.

**Review by CathPag — COMMENTED** [review]
Sorry, one more comment. Just noticed when reading the entire entry.

File: content/en/debugging.md
Comment (iamNoah1) [inline content/en/debugging.md:8]: I would replace output with result.

File: content/en/debugging.md
Comment (iamNoah1) [inline content/en/debugging.md:11]: Maybe I am too picky here, but while devs write source code, they don't necessarily create binary files. Sure, apps. run a binary at a very low level, but I think this fact does not add any value to explaining the term. Same goes  for "with external libs". 

Also it is not only about running in production. Debugging is relevant for apps running on any environment, right? 

"To identify the source of the problem, developers have to review the code line-by-line" That is not the whole truth. While it might be true, it reads as if this is the only possibility devs have. We can go through logs as well in order to find what the app is doing, without having to lock at code at all.

**Review by iamNoah1 — COMMENTED** [review]
While reading the term definition I had the impression that we are mixing up Debugging itself as a process and IDE supported Debugging. Especially the how it helps section focuses on IDE based or to be more general tool based debugging. The other sections read more like the general debugging concept or process. My suggestion is to step back and think about the scope about this term. Is it really about debugging as a process/concept? Or is it about Tool based Debugging?

**Comment by fsbaraglia** [comment]
@iamNoah1  ok, I will have a look on how to simplify. On the other hand, we should not over complicate. 

What is debugging at the end? 
A category of tools and multiple techniques to find bugs. 
Nothing more and nothing less.
Or am I wrong @CathPag  ?

**Comment by iamNoah1** [comment]
@fsbaraglia sure I am not aiming for complicating things as well :) But for correctness. Consider the following two sentences.

"A category of tools and multiple techniques to find bugs."

"Multiple techniques to find bugs, which can be tool based/assisted"

My opinion is that Debugging per se is not tools, and can be done without additional tooling at all. 

But lets see what other folks think. @JasonMorgan @seokho-son @jihoon-seo @CathPag

**Comment by CathPag** [comment]
> While reading the term definition I had the impression that we are mixing up Debugging itself as a process and IDE supported Debugging. Especially the how it helps section focuses on IDE based or to be more general tool based debugging. The other sections read more like the general debugging concept or process. My suggestion is to step back and think about the scope of this term. Is it really about debugging as a process/concept? Or is it about Tool based Debugging?

I think it should be about the concept itself

**Comment by CathPag** [comment]
> @fsbaraglia sure I am not aiming for complicating things as well :) But for correctness. Consider the following two sentences.
> 
> "A category of tools and multiple techniques to find bugs."
> 
> "Multiple techniques to find bugs, which can be tool based/assisted"
> 
> My opinion is that Debugging per se is not tools, and can be done without additional tooling at all.
> 
> But lets see what other folks think. @JasonMorgan @seokho-son @jihoon-seo @CathPag

I'll let the technical approvers chime in. My knowledge ton that subject is very limited.

**Comment by fsbaraglia** [comment]
Ok, I got you point @iamNoah1
Less tool more on concept right?

if you can please help us and make some suggestions on what to change , really appreciated

File: content/en/debugging.md
Comment (iamNoah1) [inline content/en/debugging.md:11]:
```suggestion
Software development is a complex activity which makes it nearly impossible to write code without introducing bugs. Those bugs lead to code that will likely not function as desired (undefined behavior) when executed. Depending on how critical an application is, bugs can have a huge negative impact, financially or even for human lives. Usually application code has to go through different stages or environments where it gets tested. The more critical an application is, the more accurate the testing has to be. 
```

File: content/en/debugging.md
Comment (iamNoah1) [inline content/en/debugging.md:14]:
```suggestion
When bugs appear anywhere, engineers will have to debug (which means finding and fixing) in order to decrease undesired behavior for production systems. Debugging is no easy task as engineers have to track down the source of the undesired behavior.  It not only requires knowledge about the code itself but also knowledge about the execution context at runtime. This is were different debugging techniques and tools come in handy. Analysis of logs, traces, and metrics, for instance, are used for debugging directly in production. In contrast to that, developers can use interactive the source code during the programming phase. Interactive debugging lets developers step through the code at runtime while they can analyse the related execution context. Once they have identified the source of the failure, they can correct the code and create a bug fix or patch.
```

**Review by iamNoah1 — COMMENTED** [review]
@fsbaraglia thanks for the patience. I had to think this one through and made some suggestions. I am keen to hear your opinion :=

File: content/en/debugging.md
Comment (fsbaraglia) [reply content/en/debugging.md:11]: I like it, good one @iamNoah1

**Comment by fsbaraglia** [comment]
I like it , I am fine with the new version @iamNoah1  , thank you for your support

**Review by iamNoah1 — APPROVED** [review]
@fsbaraglia thanks for your patience and contribution :)

**Review by seokho-son — APPROVED** [review]
Thanks @fsbaraglia !
LGTM.

I think this PR is ready to be merged.
To make this document live, we need to change `status: Feedback Appreciated` to `status: Completed`. 
I guess @CathPag can make a final decision ;)

**Review by CathPag — COMMENTED** [review]
I think we need to remove two words from one sentence. Please see my comment.

File: content/en/debugging.md
Comment (iamNoah1) [reply content/en/debugging.md:14]: could be my fault, as I made a little edit because some suggestions were merged resulting in a strange sentence. I 
Agree with Catherine's suggestion :)

**Comment by fsbaraglia** [comment]
> @fsbaraglia thanks for the patience. I had to think this one through and made some suggestions. I am keen to hear your opinion :=

alles klar, danke dir 💯

File: content/en/debugging.md
Comment (fsbaraglia) [reply content/en/debugging.md]: fixed :D thx for the suggestion

**Pull request merged**
